### ISSUE TITLE
allow some extension mechanism so that requests from pivot to druid can be decorated

### ISSUE THREAD

**Issue by himanshug** [implydata/pivot#138]
We are planning to turn-on authentication on our druid cluster. Druid user would be expected to provide "cookie" or a secret "token" to druid so that druid node can authenticate the user.

My understanding is that pivot sends requests to druid coordinator and broker for introspection and user queries. Many of our users use pivot and we would like to support pivot even when auth is turned on for a druid cluster.

I think it can be enabled if it is possible to decorate the requests (with custom java code preferably or else with java script code) so that we can put the necessary information in the request that pivot sends to druid.

**Comment by vogievetsky** [comment]
Sounds cool. I got the scoop on this from @gianm . 

Let me outline my idea for what I am thinking of.

In the config.yaml you would add a key of `druidRequestDecorator` that point to a local js file 

```
druidRequestDecorator: ./himanshu_decorator.js
```

Then the contract is that from that file you export a function `decoratorFactory` that has to return a decorator which would just be a function that gets called on every request and gets the druid query and can return an object with the key `headers` where you can set whatever headers you want.

himanshu_decorator.js might look like this:

```javascript
var some_fancy_module = require('some_module_that_you_npm_installed');
var leftPad = require('left-pad');

exports.decoratorFactory = function(decoratorFactoryOptions) {
  var someToken = leftPad('get_your_token_here', '$', 100);  
  // you could even call an external process

  return function(druidQuery, decoratorParameters) {
    var someCustomHeaders = {
     'X-Y-AUTH': someToken
    };

    return {
      headers: someCustomHeaders
    };
  };
};
```

Does that sound like it would work? Do you need any specific info passed in as decoratorFactoryOptions or as decoratorParameters?

There is nothing that Pivot knows that you can not know ahed of time and encode it directly into your code so I think there is not too much helpful stuff to add there.

**Comment by himanshug** [comment]
it should work for the current scenario we have 
but it would be nice if druidQuery was not just the query but full request object obtained from user/browser. I can imagine ppl trying to put some code in the extension to authenticate requests coming to pivot itself based on some headers or query params etc in the request and not just decorating outgoing requests to druid.

**Comment by himanshug** [comment]
also, what is in the "decoratorParameters? ?

**Comment by vogievetsky** [comment]
decoratorParameters is something that Pivot passes to you that will have cool stuff in it maybe the full request object as you said. decoratorFactoryOptions would contain the config that pivot parsed.

**Comment by vogievetsky** [comment]
This is in PR now: https://github.com/implydata/pivot/pull/141/files

**Comment by vogievetsky** [comment]
New Pivot published (0.8.2) supports request decoration. You can read more about it here: https://github.com/implydata/pivot/blob/master/docs/security.md#authentication I am going to close this issue but please keep commenting here or start a thread on https://groups.google.com/forum/#!forum/imply-user-group if you encounter any difficulties.
